### PR TITLE
support user properties

### PR DESCRIPTION
**Pull request by follitude**

### CODE REVIEW

File: librocksdb_sys/crocksdb/c.cc [hunk 1]
```diff
@@ -120,6 +128,7 @@ struct crocksdb_livefiles_t       { std::vector<LiveFileMetaData> rep; };
 struct crocksdb_column_family_handle_t  { ColumnFamilyHandle* rep; };
 struct crocksdb_envoptions_t      { EnvOptions        rep; };
 struct crocksdb_ingestexternalfileoptions_t  { IngestExternalFileOptions rep; };
+struct crocksdb_tablepropertiescollectorfactorycontext_t {TablePropertiesCollectorFactory::Context rep;};
```
Comment (andelf) [inline librocksdb_sys/crocksdb/c.cc]: remove

File: librocksdb_sys/src/lib.rs [hunk 2]
```diff
@@ -45,6 +45,9 @@ pub enum DBRateLimiter {}
 pub enum DBLogger {}
 pub enum DBCompactOptions {}
 pub enum DBPinnableSlice {}
+pub enum DBTablePropertiesCollector {}
+pub enum DBTablePropertiesCollectorFactory {}
+pub enum DBTablePropertiesCollectorFactoryContext {}
```
Comment (andelf) [inline librocksdb_sys/src/lib.rs]: not used

File: src/table_properties_collector.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                             value.len());
+            *(*values.offset(i as isize)).offset(value.len() as isize) = 0;
+        }
+        *c_keys = keys;
+        *c_values = values;
+        *pair_count = count as c_int;
+    }
+}
+
+pub extern "C" fn readable_properties(collector: *mut c_void) {
```
Comment (andelf) [inline src/table_properties_collector.rs]: not used now

Comment (andelf) [inline src/table_properties_collector.rs]: better add TODO: or sth

**Comment by andelf** [comment]
Run clang-format and cargo fmt before make a PR.
:)

File: tests/test_table_properties_collector_factory.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    let db = DB::open(opts, path.path().to_str().unwrap()).unwrap();
+    let samples = vec![(b"key1".to_vec(), b"value1".to_vec()),
+                       (b"key2".to_vec(), b"value2".to_vec())];
+    for &(ref k, ref v) in &samples {
+        db.put(k, v).unwrap();
+        assert_eq!(v.as_slice(), &*db.get(k).unwrap().unwrap());
+    }
+    db.flush(true).unwrap();
+    drop(db);
+    println!("path is {:?}", path);
```
Comment (siddontang) [inline tests/test_table_properties_collector_factory.rs]: remove it

File: tests/test_table_properties_collector_factory.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+use rocksdb::crocksdb_ffi::EntryType;
+use std::collections::HashMap;
+use tempdir::TempDir;
+
+struct CollectorFactory {
+    _cnt: u64,
+}
+
+struct Collector {
+    cnt: u32,
```
Comment (siddontang) [inline tests/test_table_properties_collector_factory.rs]: where to check cnt in test?

**Comment by siddontang** [comment]
Please fix CI

**Pull request closed without merging**